### ISSUE TITLE
ExtensionArray.map

### ISSUE THREAD

**Issue by TomAugspurger** [pandas-dev/pandas#23179]
Both Categorical and SparseArray found implementing a `.map` method useful. This allows them to efficiently apply a function / mapping to the categories / sp_values, rather than every element of an array. We dispatch to it internally in https://github.com/pandas-dev/pandas/blob/master/pandas/core/series.py#L3379-L3380

So, we need to either

1. Add it to the interface
2. hard-code checks for categorical or sparse dtype there.

Do people have a preference? Right now I'm leaning toward 2. Or are there other array types that would have a similar efficiency gain to Categorical or Sparse?

**Comment by jreback** [comment]
-1 on hard coding things

expanding the interface is the way forward here

**Comment by jbrockmendel** [comment]
Trying to de-duplicate is_extension_array_dtype and is_extension_type, I'm finding that the lack of EA.map is a blocker for using is_extension_array_dtype in all cases.

**Comment by TomAugspurger** [comment]
FWIW, I don't think deduplicating is_extension_array_dtype
and is_extension_dtype important enough to warrant adding a new method to
the API.

**Comment by rhshadrach** [comment]
Ran into this in #39941, where `map` is used for categorical and sparse in `apply`. Here, it results in different dtype behavior than other EAs. But it seems to me that `map` only makes sense when any UDF can be remain in the same dtype (which I _think_ is true for categorical and sparse). But how would one implement map for e.g. `Int64` where the mapper is `lambda x: 3.2` or `lambda x: "a"`?

**Comment by jbrockmendel** [comment]
> any UDF can be remain in the same dtype (which I think is true for categorical and sparse)

Not exactly.  For sparse you can make the return dtype always be sparse, but we can come up with UDFs that must have different sparse dtype.  For Categorical you _could_ pass your result to `type(self)._from_sequence(result, dtype=self.dtype)` and that will usually work, but thats bc it will just set any non-fitting element to nan.

**Comment by topper-123** [comment]
Closed. `ExtensionArray.map` was added in #51809.
